stack: Submission of the Haskell Tool Stack

essandess opened the pull request:
stack: Submission of the Haskell Tool Stack

* This runs the standard stack installation into ${prefix}/bin
* A bootstrap variant is included for stack to build itself

#### Description

<!-- Note: it is best make pull requests from a branch rather than from master -->

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [x] submission
- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on
<!-- Generate version information with this command in shell:
    echo "macOS $(sw_vers -productVersion) $(sw_vers -buildVersion)"; echo "Xcode $(xcodebuild -version | awk '{print $NF}' | tr '\n' ' ')"
-->
macOS 10.14.5 18F132
Xcode 10.2.1 10E1001

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?
<!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [x] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

pmetzger commented on lang/stack/Portfile:
```diff
@@ -0,0 +1,62 @@
+# -*- coding: utf-8; mode: tcl; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- vim:fenc=utf-8:ft=tcl:et:sw=4:ts=4:sts=4
+
+PortSystem          1.0
+PortGroup           github                      1.0
```
There's no need for all the space between `github` and `1.0`

essandess replied:
Deleted, thanks.

pmetzger commented on lang/stack/Portfile:
```diff
@@ -0,0 +1,62 @@
+
+name                stack
+categories          lang haskell
+maintainers         nomaintainer
```
You don't want to be the maintainer?

essandess replied:
I’ll be the maintainer.

reneeotten replied:
the maintainer line should follow: ```{example.org:julesverne @github-handle} openmaintainer``` and unless there is a good reason not to, adding "openmaintainer" to allow others to make "minor changes" without your approval would be preferred

essandess replied:
Done. Thanks.

pmetzger commented on lang/stack/Portfile:
```diff
@@ -0,0 +1,62 @@
+description The Haskell Tool Stack
+long_description    ${description}. \
+    Stack is a cross-platform program for developing Haskell projects. \
+    It is intended for Haskellers both new and experienced.
```
When you say "program", what do you mean? IDE? (Given that I can't tell, the description may need some work.)

essandess replied:
Copy-and-paste description text from their main page. I edited this a little.

pmetzger commented on lang/stack/Portfile:
```diff
@@ -0,0 +1,62 @@
+} else {
+    fetch {}
+    checksum {}
+    extract.cmd /usr/bin/true
```
I don't quite get this bit of logic. You might want to comment it. (Indeed, you might want to comment the whole bootstrap thing, especially if you're not maintaining this, so others can understand it.)

essandess replied:
Now new logic based on your comments below and aversion to `curl | sh`.

The logic now is:
1. Download the pre-built binary from the github release page, then checksum.
2. Destroot the pre-built binary.
3. If bootstrapping, overwrite the pre-built binary.

pmetzger commented on lang/stack/Portfile:
```diff
@@ -0,0 +1,62 @@
+# stack must bootstrap itself
+# See: https://github.com/commercialhaskell/stack/blob/master/doc/README.md
+pre-build {
+    system -W ${worksrcpath} "/bin/mkdir ./bin && /usr/bin/curl -sSL https://get.haskellstack.org/ | /bin/sh -s - -d ./bin"
```
Doing curl into sh without checksumming what's fetched seems very bad. I think normal users would not expect something like that to occur in macports.

essandess replied:
Yeah, that’s what the standard install is, but still…

I changed this to a direct download from the GitHub release page with a checksum.

reneeotten commented on lang/stack/Portfile:
```diff
@@ -0,0 +1,62 @@
+        building your project, testing your project, benchmarking your project. \
+        It is intended for Haskellers both new and experienced.
+
+homepage            http://haskellstack.org
```
can we use https?

essandess replied:
Yes, done, thx.

reneeotten commented on lang/stack/Portfile:
```diff
@@ -0,0 +1,62 @@
+	    description {Bootstrap a build of stack.} {}
+
+master_sites        ${github.homepage}/releases/download/v${github.version}
+distname            ${name}-${github.version}-osx-x86_64
```
given that you're getting a file for "x86_64", you probably want to add ```supported_archs     x86_64``` (see the [guide](https://guide.macports.org/#reference.keywords))

essandess replied:
Yes. Done, thx.

essandess replied:
Done, thx.

essandess commented:
@pmetzger @reneeotten Thanks for your comments and corrections.

A couple questions:
1. I believe that this port `stack` should supplant the current method of providing `ghc`—the Portfile for `ghc` Is way back in version 7 and unusably out-of-date. Is there a way of looping in its maintainer `{cal @neverpanic}` ?
2. Is there any practical reason for bootstrapping `stack` or `ghc` within MacPorts itself? Both are self-compiling and the binaries must necessarily come from outside MacPorts. It’s a lot of work to maintain a self-licking ice cream cone, and would be good to avoid if necessary.

pmetzger commented:
If our ghc is way out of date, the fastest thing to do is submit an update for that. If you want to loop someone in, just @-mention them. (Putting their ID in backquotes prevents them from being alerted.)

mojca commented:
> If our ghc is way out of date, the fastest thing to do is submit an update for that.

Except that updating it is **anything** but trivial. I tried twice and then gave up before I managed to finish it. This might need to be done in close collaboration with upstream developers advising about how to proceed.

pmetzger commented:
Generally the deal with haskell, ocaml, and a bunch of other similar ecosystems is that they're really intended to be built entirely with their own tools inside their own package system. This makes dealing with them unpleasant because there are end applications MacPorts needs to be able to build that use those compilers, but the maintainers don't think in terms of anyone trying to use the compiler outside of their own package manager.

essandess commented:
I’ve found the stack binaries are robust within its own walled ecosystem—self-updating works, builds work, ghc installs all work.

But as @mojca says, it’s a major pain and time-sink to try to rebuild the binaries within MacPorts (or elsewhere).

My suggestion would be to use this stack port as a MacPorts-flavored stand-in to the `curl | sh` binary install, then just use stack to update itself, install ghc, and everything else stack does. That would be a **lot** simpler and easier to maintain than the current port of `ghc`.

I’m trying to provide the basic Portfile hooks for a bootstrap of stack, but this still doesn’t compile, and I doubt it’s worth the effort just to re-create the working, supplied binary.

essandess commented:
What’s the MacPorts Way for verifying `gpg` signing keys. The stack binaries are all signed with https://docs.haskellstack.org/en/stable/SIGNING_KEY/.

Add `port:gnupg` as a build_depends, then a bunch of system commands to install the key in root’s keychain and verify the binary? Is there a representative copy-and-paste Portfile example?

mojca commented:
It would be nice if @neverpanic commented about gpg and ghc in general.

But one general rule (maybe one security step lower than gpg) is that we always fetch a set of files in advance and checksum them all. Fetching the files during the build is not desired / allowed. The file which you are fetching may not change over time.

essandess commented:
Also note that stack’s macOS install instructions include `brew` with these notes:
* https://docs.haskellstack.org/en/stable/install_and_upgrade/
* https://github.com/Homebrew/brew/blob/master/docs/FAQ.md#why-do-you-compile-everything
* https://github.com/Homebrew/formulae.brew.sh/blob/master/_data/formula/haskell-stack.json

The current brew version of stack is pre-2.0, and I’m not conversant enough with bottles or casks to figure out if brew compiles from source, or how they download the original binary.

mojca commented:
Homebrew tends to install a lot of things from precompiled binaries. But you should be looking at the ruby code where you can see what precisely they download:
* https://github.com/Homebrew/homebrew-core/blob/master/Formula/ghc.rb
* https://github.com/Homebrew/homebrew-core/blob/master/Formula/haskell-stack.rb

The downloaded files are these, for example (or even the git directory):
* https://downloads.haskell.org/~ghc/8.6.5/ghc-8.6.5-src.tar.xz
* https://github.com/commercialhaskell/stack/releases/download/v1.9.3/stack-1.9.3-sdist-1.tar.gz

pmetzger commented:
Generally: MacPorts has a bias towards doing our own builds (which assures we know what sources went into binaries we distribute) and towards careful technical decisions. How this fits in with our decisions on the Haskell platform, though, isn't clear.

neverpanic commented:
TBH, it's quite obvious that I don't have the time to deal with packaging GHC, so I'm fine with whatever solution solves the stalemate we currently have. I'd prefer if we actually bootstrapped our own stack, but the tooling seems to be too complicated to maintain this within macports.

Maybe the proper solution would be to package only stack and then have users use it to install the tools they want into their home directories? Using stack to manage files inside the macports prefix is not a good solution and asking for trouble – macports' database won't know about the files, will not uninstall them and will cause problems when other ports attempt to write to the same files.

pmetzger commented:
> Using stack to manage files inside the macports prefix is not a good solution and asking for trouble – macports' database won't know about the files, will not uninstall them and will cause problems when other ports attempt to write to the same files.

+1

essandess commented:
It’s more nuanced than that. `stack` can be used within a Portfile to destroot MacPorts-managed files like any other tool.

But when used to build a project, and set by the user as a shell variable, it will download and add a bunch of files to `STACK_ROOT`, which I’ve provided as `${prefix}/etc/.stack` in this submission. Perhaps we shouldn’t do that. Is there a clean way if blowing away the entirety of `STACK_ROOT` when/if the port stack is uninstalled?

essandess commented:
I refactored the Portfile a little based on this discussion.
* Independent checksum test
* No gpg test because the signature is signed with a key that doesn’t match the advertised key (!)
* No creation of `${prefix}/etc/.stack`
* A bootstrap build that should work but doesn’t because of some missing hackage tarball—see https://paste.macports.org/69398e9e9981

Here’s an example suggested usage of the stack port: https://github.com/macports/macports-ports/pull/4632. This would be the basic approach for porting the latest `ghc`, bootstrapped, if that works…

I’ll turn off default bootstrapping and believe this is a ready to go if you all approve.

essandess commented:
The refactoring of [port:pandoc](https://github.com/macports/macports-ports/pull/4698) is a good illustration of how stack can simplify Haskell builds and project management.

If MacPorts is used as a package-manager-for-a-package-manager, like the Portfile for `pandoc` with over two dozen (!) `hs-*` out-of-date port dependencies, the project become unmanageably complicated and effectively impossible to update.

The build for `pandoc` uses stack [explicitly](https://github.com/jgm/pandoc/blob/master/stack.yaml), so it’s only necessary to download the source, run stack, and let stack take care of all the source dependencies and versioning.

The `pandoc` Portfile at #4698 will update to the latest version, and will be trivial to update when the next version comes along. It will be the same for any Haskell project that’s configured to build with stack.

neverpanic commented on lang/stack/Portfile:
```diff
@@ -0,0 +1,99 @@
+
+pre-extract {
+    # second SHA256 comparison against stack-provided hash
+    set sha256_verification [exec /bin/sh -c "\[\[ \$(< ${distpath}/${name}-${github.version}-osx-x86_64${extract.suffix}.sha256) == \"${sha256_checksum}\" \]\] && echo 'VERIFIED' || echo 'UNVERIFIED'"]
```
What are you trying to achieve with this second checksum? Isn't the downloaded file already checksummed?

essandess replied:
I originally was verifying with gpg key, but the upstream distro used an inconsistent signature (!), so I’m using the distribution-provided checksum instead, which should be the same as the MacPorts-provided checksum, but does provide another independent check for this non-bootstrapped binary.

neverpanic replied:
That's my point – I don't think this actually gives us any additional assurance from a security PoV, and I don't really like using shell scripts inside Tcl, so I'd prefer if we just removed this check entirely.

essandess replied:
Okay, I'll delete it and add in gpg verification if stack fixes their signatures.

essandess commented:
Bootstrapping now works. After a successful Travis or Azure build I’d say this is good to go.

essandess commented:
@pmetzger This is ready to commit, and is in the critical path of other pending submissions.

mf2k commented:
Error is:

```
DEBUG: system -W /opt/local/var/macports/build/_Users_vsts_agent_2.154.1_work_1_s_lang_stack/stack/work: /bin/sh -c '/opt/local/bin/gpg --homedir /opt/local/var/macports/build/_Users_vsts_agent_2.154.1_work_1_s_lang_stack/stack/work/.gnupg --import /Users/vsts/agent/2.154.1/work/1/s/lang/stack/files/keyid-1c5c154d.txt 2>/dev/null || /usr/bin/true'
Error: Failed to checksum stack: can't read "keyid": no such variable
```

essandess commented:
This is ready to commit.

neverpanic commented on lang/stack/Portfile:
```diff
@@ -0,0 +1,116 @@
+
+supported_archs     x86_64
+
+post-checksum {
```
Same comment as for the GHC Portfile: I really think the maintainer should do this and the users should rely on the checksums in the Portfile that already provide a chain of trust, rather than starting multiple complex shell commands here.

essandess replied:
Thanks. If there's consensus not to check author PGP signatures within MacPorts, I'll remove these blocks.

Any other downvotes for signature checking? I'll count @neverpanic and @pmetzger as recommending removal.

essandess replied:
It's gone. Also see these comments: https://github.com/macports/macports-ports/pull/4825#discussion_r304193581

essandess commented:
@mf2k @pmetzger @neverpanic @mojca @cjones051073 

This Portfile is working great, and there’s no more tests or bells to add.

Would you all please take a look and merge if you agree? It’s in the critical path of a couple other PRs.

essandess commented:
@neverpanic Any more comments or remaining issues? May we please merge this?

I'm hoping to use this port to update `pandoc` (https://github.com/macports/macports-ports/pull/4698) from version 1.12—**released over five years ago on [May 8, 2014](https://github.com/jgm/pandoc/releases?after=1.13.1)**—to the latest version 2.7 released last month.

Also, I've seen firsthand that the buildbot build of panda often results in Travis errors. E.g. see https://paste.macports.org/aee868603e51 while building the dependencies for https://github.com/macports/macports-ports/pull/4914:
```
Error: Failed to fetch hs-quickcheck: Could not resolve host: hackage.haskell.org
DEBUG: Error code: NONE
DEBUG: Backtrace: Could not resolve host: hackage.haskell.org
```

I'm hoping that using modern Haskell development tools and code will make Travis builds more reliable.

essandess commented:
@neverpanic @mf2k @pmetzger @mojca @cjones051073 

This port works. I’ve had the stack binaries deployed for over a month without issue.

May we please merge or identify any remaining issues?

essandess commented:
@mf2k I implemented your suggested mods and updated everything to the current stack version 2.1.3. BTW `master_sites` is still required to work (otherwise it does a clone).

Once this builds, its good to merge.

essandess commented:
@mf2k This is good to merge.

mf2k commented:
@pmetzger: Waiting on your change request approval.

cjones051073 commented:
@pmetzger ping ?

essandess commented:
@pmetzger

May we please merge this so it can be used to update other out-of-date ports?

cjones051073 commented:
@pmetzger needs to agree to the changes before this can be merged.

neverpanic commented:
It seems this port requires internet access during the build phase and downloads another copy of GHC (8.2.2 to be specific). Is that really what we want as the standard packaging mechanism for all our haskell ports?

Also, you indicated that you tried this build with the `-vst` flags, but for me it fails with `-t` (trace mode) enabled, because it seems to require access to a certificate store, and trace mode seems to deny that. Did this work for you?

neverpanic commented:
Also, shouldn't we do the same thing the brew people do (https://github.com/Homebrew/homebrew-core/blob/master/Formula/haskell-stack.rb) and pass `--no-install-ghc --system-ghc` as well as passing `--resolver lts-14.1` to get https://www.stackage.org/lts-14.1?

essandess commented:
> It seems this port requires internet access during the build phase and downloads another copy of GHC (8.2.2 to be specific). Is that really what we want as the standard packaging mechanism for all our haskell ports?

I’d say yes—strongly. Each Haskell package built with `stack` will specify—explicitly or implicitly—the `ghc` version required to get it to build. The whole point of `stack` is to go and download the required `ghc` version along with the correct version of all the dependencies. It’s completely unnecessary as well as a maintenance nightmare for MacPorts to duplicate all that infrastructure. Just let `stack` take care of it.

Think of the MacPorts `ghc` port as an up-to-date Haskell compiler that may sometimes be used to build specific ports (like `cabal`), but is mostly just there to run `ghci`. The development tools `stack` and `cabal` will grab the necessary versions. 

Without online access, neither `stack` nor `cabal` will operate as designed, or as the package developer that designs their package for builds using these tools intends.

> Also, you indicated that you tried this build with the `-vst` flags, but for me it fails with `-t` (trace mode) enabled, because it seems to require access to a certificate store, and trace mode seems to deny that. Did this work for you?

Not sure anymore…

essandess commented:
> Also, shouldn't we do the same thing the brew people do (https://github.com/Homebrew/homebrew-core/blob/master/Formula/haskell-stack.rb) and pass `--no-install-ghc --system-ghc` as well as passing `--resolver lts-14.1` to get https://www.stackage.org/lts-14.1?

I’d say no—my own preference would be to build it [exactly the way they say it should be built](https://github.com/commercialhaskell/stack/blob/master/snapshot.yaml), which currently is `lts -11.22` and `ghc-8.2.2`. That way, if something ever breaks it’s a lot easier to get support by posting an issue that says “you told me to do this thing and I did that thing and it’s broken,” rather than trying to second guess the upstream build decisions and being told “you didn’t do that thing I told you to do. Go do that thing!”

neverpanic commented:
> Think of the MacPorts `ghc` port as an up-to-date Haskell compiler that may sometimes be used to build specific ports (like `cabal`), but is mostly just there to run `ghci`. The development tools `stack` and `cabal` will grab the necessary versions.

OK, but that will mean that we trust whoever compiled those binaries (an argument that's not really very good in the Haskell world anyway, since we need a precompiled binary compiler to bootstrap GHC itself as well).

> Without online access, neither `stack` nor `cabal` will operate as designed, or as the package developer that designs their package for builds using these tools intends.

So far, our Haskell packages used cabal and the downloading was done by MacPorts (which allows us to mirror, cache and verify the downloads). This would no longer be the case with stack.

essandess commented:
> OK, but that will mean that we trust whoever compiled those binaries (an argument that's not really very good in the Haskell world anyway, since we need a precompiled binary compiler to bootstrap GHC itself as well).

Yep. 

> So far, our Haskell packages used cabal and the downloading was done by MacPorts (which allows us to mirror, cache and verify the downloads). This would no longer be the case with stack.

Yep. We’ll be relying on `stack` to take care of all the download verification, like everyone else who uses Haskell packages built with `stack` (the majority now, I believe). Stack design [explicitly](https://www.fpcomplete.com/blog/2015/06/why-is-stack-not-cabal) relies upon this online stack of stuff to work:
* Git (for package index management),
* S3 (for high-reliability package serving),
* SSL libraries (for secure HTTP uploads and downloads),
* Docker,
* many state-of-the-art Haskell libraries.
All the downloading and hash checking based on the `stack.yaml` or `.cabal` Configuration is done in ` stack`’s [setup](https://github.com/commercialhaskell/stack/blob/master/src/Stack/Setup.hs) code.

@borsboom or @snoyberg could comment much more specifically about stack’s security model and what they do to get trusted `ghc` and code downloads.

Practically, MacPorts cannot replicate or refactor Haskell world’s 20+ years of combined community effort, experience, and false starts in the tooling that exists in `cabal` and `stack`. Also, current MacPorts’s Haskell packages don’t use `cabal` for builds (see https://github.com/macports/macports-ports/search?p=2&q=cabal&unscoped_q=cabal).

If we want to build Haskell packages that specify `stack` to build them, we’ll need `stack`, and the design features that go with it.

essandess commented:
@neverpanic The dependency chain for downloading a trusted `ghc` and hackage code is defined by the constant `defaultSetupInfoYaml` = https://raw.githubusercontent.com/fpco/stackage-content/master/stack/stack-setup-2.yaml in [Setup.hs](https://github.com/commercialhaskell/stack/blob/master/src/Stack/Setup.hs). This hardcoded default setup points to, e.g., these ghc 8.6.5 links for macOS:
```
        8.6.5:
            url: "https://github.com/commercialhaskell/ghc/releases/download/ghc-8.6.5-release/ghc-8.6.5-x86_64-apple-darwin.tar.bz2"
            content-length: 226259652
            sha1: f15d2dac138e391f0c3a1657e9cfd8cd69fd6c8b
            sha256: 1cc7c7bd7ccd5a7ac7fdd2f6913c070694bf974c9f2fd0df9fe41c7babdab9fd
```

Two hash and a length check are performed for verification, just like MacPorts port files. An attack would require compromise of either [FP Complete](https://github.com/fpco)’s or [commercialhaskel](https://github.com/commercialhaskell)’s GitHub accounts, an open source attack at https://hackage.haskell.org/ itself, or a malicious package.

As you point out, just by using `stack` (online or off), we’re already on these attack surfaces.  We’re using MacPorts to download, verify, and build trusted code (`stack`) from @commercialhaskel, which is then run to download, verify, and build trusted constants from @fpco and code from @commercialhaskel and hackage.

essandess commented:
@pmetzger (@cjones051073, @mf2k) Now that the `ghc` port is updated to the latest, the currently very-very-out-of-date port `pandoc` won’t build. I’ve fixed this with an upgrade to the latest `pandoc`  in PR https://github.com/macports/macports-ports/pull/4698.

Many ports have `pandoc` as a dependency, so this breakage will propagate.

Would you please merge the PR so that it can be used for other builds, or identify any outstanding issues?

essandess commented:
I simplified the (unused) gpg checking by removing unnecessary hash checking (and keeping length checking because checksum.tcl says at least one check is necessary) in anticipation of when this gets implemented in a portgroup or something.

@pmetzger this is ready to merge.

essandess commented:
@mf2k @cjones051073 @neverpanic, @pmetzger says in PR https://github.com/macports/macports-ports/pull/4698#issuecomment-524624642 that he doesn't have the time to approve these changes.

Would one of you please merge this port as soon as possible so that we can move forward fixing other Haskell stuff?

Many existing Haskell ports are broken and years out of date and need modern tools to fix them now that `ghc` has been updated. See https://github.com/macports/macports-ports/pull/5050.

mojca commented:
> Would one of you please merge this port as soon as possible so that we can move forward fixing other Haskell stuff?
> 
> Many existing Haskell ports are broken and years out of date and need modern tools to fix them now that `ghc` has been updated. See #5050.

A note for the future (now it's too late already anyway): when one PR depends on another, and a bunch of other stuff also depends on both and breaks as a side effect in the meantime, it's a good idea to put all those changes to the same pull request (so that everything gets updated at the same time), and try to put the best effort into not breaking anything in the process.

I admit that it was / is very non-trivial to do everything right for Haskell, but let's make the best effort not to break dependencies in the future (and fix the remaining broken ports ASAP).

mojca commented:
Sadly this is broken on our < 10.9 builders due to lack of support for newer encryption standards, and the workarounds we use don't really help as stuff is being downloaded on the fly. (https://ports.macports.org/port/stack/)

ryandesign commented:
> Sadly this is broken on our < 10.9 builders due to lack of support for newer encryption standards, and the workarounds we use don't really help as stuff is being downloaded on the fly.

This is now filed as https://trac.macports.org/ticket/58934.
